docs: add doc with all the steps required for argocd-agent with argocd-operator (#482)
argoproj-labs/argocd-agent: docs/getting-started/openshift/index.md

# ArgoCD Agent Hub and Spoke Cluster Setup Documentation Using Operator
This document outlines the process of setting up ArgoCD Agent in a hub and spoke cluster architecture. This configuration allows a central ArgoCD instance (the hub) to manage applications deployed across multiple remote Kubernetes clusters (the spokes) through the argocd-agent.
## Prerequisites

> Before proceeding with the setup, ensure you have the following:

Reviewer: Would be nice to have bullet points for prerequisites

A running openshift/kubernetes cluster designated as the hub, with ArgoCD/openshift GitOps Operator installed.
One or more Kubernetes clusters designated as spokes, ArgoCD/openshift GitOps Operator installed.
oc configured to access both hub and spoke clusters.

> Operator must be installed in cluster scope mode.

Reviewer 2: The Argo CD instance must be in cluster scoped mode right ? Operator is by default in cluster scoped (installs in All Namespaces)

Reviewer: I was trying out autonomous mode without making it cluster scoped. Seemed to be working. I need to try it out again on the new staging build for confirming because previously I was facing issues with createNamespace not working and I didn't try it out after

Reviewer 2: can it be reworded ? instead of saying `Operator` mention `Argo CD instance must be cluster scoped`.

Author: update

> Apps in any Namespace should be enabled.

Reviewer 2: `Apps in any Namespace` feature must be enabled in the Hub cluster.

Reviewer: This doesn't seem like a requirement for autonomous mode. I guess it would depend on how the customers would like to use it

Author: Should I mention explicitly for managed?

Reviewer 3: It will be required for autonomous mode too, because the principal will replicate Applications in a namespace with the same name as the agent that created them.

Argocd-agentctl binary (for non-production scenario)

> 

Reviewer 2: `argocd-agentctl` binary must be installed.

Author: But we are not recommending it for production use, so should we share it as a must in documentation?

Reviewer 3: If you use `argocd-agentctl`, the `openssl` binary is not required

Author: Should we mentaion either openssl or argocd-agentctl as a requirement?

## Hub Cluster Setup

Hub cluster will be installed in argocd namespace

> Before you install the principal's manifest, you will need to create three

Reviewer: This sentence seems incomplete. Should we mention Principal component before this? Its referred to as "hub" in general in this document.
Also, would be nice to have this in bullets

Reviewer 3: "Principal" and "hub" are two very different entities. Please refer to https://argocd-agent.readthedocs.io/latest/concepts/components-terminology/

```
argocd-agentctl pki issue principal \
    --principal-context <control plane context> \
    --ip <ip addresses of principal> \
    --dns <dns names of principal>
```

For the --ip and --dns values, you want to specify all addresses and DNS names that the principal will be reachable at for the agents, for example --ip 5.5.5.5 --dns my-principal.example.com. If there is a load balancer in front of the principal, make sure it will pass through traffic to the principal - otherwise, client certificates might not be used for authentication.

Issue the certificate for the principal's resource proxy:
The principal's resource proxy needs to be reachable by your Argo CD API server (argocd-server), which usually runs in the same cluster as the principal. Typically, this service should not be reachable outside the cluster and is exposed using a Kubernetes Service.

```
argocd-agentctl pki issue resource-proxy \
    --principal-context <control plane context> \
    --ip <ip addresses of resource proxy> \
    --dns <dns names of principal>
```

Create argocd-redis secret, because principal looks for it to fetch redis authentication details.

```

> oc create secret generic argocd-redis -n argocd --from-literal=auth="$(oc get secret    argocd-redis-initial-password -n argocd -o jsonpath='{.data.admin\.password}' | base64 -d)"

Reviewer: ```suggestion
oc create secret generic argocd-redis -n argocd --from-literal=auth="$(oc get secret argocd-redis-initial-password -n argocd -o jsonpath='{.data.admin\.password}' | base64 -d)"
```

Reviewer: Can you change the order? This needs to happen after ArgoCD CR creation

Author: sure

```

Deploy agent on hub cluster using argocd-operator/gitops-operator using ArgoCD CR given below

```
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd
  Namespace: argocd
spec:
  controller:
    enabled: false
  argoCDAgent:
    principal:
      enabled: true
      allowedNamespaces: 
        - "*"
      jwtAllowGenerate: true
      auth: "mtls:CN=([^,]+)"
      logLevel: "trace"

>       image: "quay.io/user/argocd-agent:v1"

Reviewer: ```suggestion
      image: "<quay.io/user/argocd-agent:v1>"
```

Should we add a note how the image would be created? For gitops customers, we'd need to share the registry image

Reviewer: Should we mention the official image now that we have release 1.17?

Author: updated

Author: > Should we mention registry.redhat.com image for gitops customers? https://catalog.redhat.com/software/containers/openshift-gitops-1/argocd-agent-rhel8/686ba1ec8da608b2007dfe1c?container-tabs=gti

I think its upstream, so I would say let's use the upstream image.

  sourceNamespaces:
    - "agent-managed"
    - "agent-autonomous"  					
```

> The above CR should create all the necessary resource for argocd as well as argocd-agent principal in argocd namespace.

Reviewer: I think we can mention metrics service as well

## Setting up agent workload cluster 

### Configure Argo-CD for Agent 
 

> argocd instance on Agent cluster

Reviewer 2: `ArgoCD` instance on agent cluster.

> Creating argocd instance for Workload/spoke cluster, this instance will be using the shared repo, redis instance from principal and, have locally running argocd-application-controller instance.

Reviewer 3: There are no shared components between principal and agent anymore since v0.2.0. The agent is supposed to use local Redis and Repository Server now.

Update the below manifest with loadbalancer ip to create argocd instance. 
```
  apiVersion: argoproj.io/v1beta1
  kind: ArgoCD
  metadata:
    name: argocd

>     namespace: argocd

Author: No, it's not required to run agent on argocd namespace, but currently with the argocd-agentctl we can use only argocd ns due to a bug, so I decided to keep it.

>     namespace: argocd

Reviewer 3: What bug?

>     namespace: argocd

Author: I have tried again, and it's working, I think I might have missed the flag on `argocd-agentctl init` command or didn't specify the namespace correctly.  I will update according to above comment.

>     namespace: argocd

Reviewer 3: It's indeed a bug, and I've submitted a fix for it: https://github.com/argoproj-labs/argocd-agent/pull/483

Please do create a bug (or submit a PR) next time you stumble over something like this.

  spec:
    repo:

>       remote: <repo LoadBalancer IP addr>:8081

Reviewer: I think we do have this by default, didn't have to make changes. For redis, we did

    redis:

>       remote: <redisLoadBalancer IP addr>:6379

Reviewer: We need to change the .spec.type to use LoadBalancer instead of ClusterIP. Worth mentioning

>       remote: <redisLoadBalancer IP addr>:6379

Reviewer 3: We do not need argocd-redis to be of type LoadBalancer

    server:
      enabled: false
    sourceNamespaces:

>     - <workload namespace>

Reviewer 3: Agent doesn't handle multiple namespaces. It's best to have the agent installed in the same namespace as the application controller, that way, you won't need sourceNamespaces.

```

### Configure Agent in managed mode

Before installing agent resources create 
Create a TLS secret containing the issued certificate for agent

Create the PKI on the agent:
Run this command while connected to principal
```
argocd-agentctl pki issue agent <agent-name> --agent-context <workload context> --agent-namespace <workload namespace> --upsert
```

Apply the installation manifests for argocd agent
```
oc apply -n $(workload-namespace) -k 'https://github.com/argoproj-labs/argocd-agent/install/kubernetes/agent?ref=main'
```
This should create all the required agent related resources.

Update the configMap with name `argocd-agent-params`  with parameters related to agent.mode,agent.creds, agent.namespace, agent.server.address.	
```

>   agent.mode: autonomous

Author: Typo while updating CM data for managed.

>   agent.creds: mtls:any

Reviewer: A little different from what I used while testing:

```
data:
  agent.keep-alive-ping-interval: 50s # Otherwise the connection would close I guess
  agent.tls.client.insecure: 'true' # This is for testing purpose
  agent.server.port: '443'
  agent.tls.root-ca-path: ''
  agent.tls.client.cert-path: ''
  agent.tls.root-ca-secret-name: argocd-agent-ca
  agent.tls.secret-name: argocd-agent-client-tls
  agent.mode: autonomous
  agent.log.level: info
  agent.namespace: argocd
  agent.server.address: <principal server address>
  agent.metrics.port: '8181'
  agent.tls.client.key-path: ''
  agent.healthz.port: '8002'
  agent.creds: 'mtls:any'
```

Author: ```
  agent.healthz.port: '8002'    <------ this was missing
```

### Configure Agent in Autonomous mode

Before installing agent resources create 
Create a TLS secret containing the issued certificate for agent

Create the PKI on the agent:
Run this command while connected to principal
```
argocd-agentctl pki issue agent <agent-name> --agent-context <workload context> --agent-namespace argocd --upsert
```

Apply the installation manifests for argocd agent
```
oc apply -n argocd -k 'https://github.com/argoproj-labs/argocd-agent/install/kubernetes/agent?ref=main'
```
This should create all the required agent related resources.

Update the configMap with name `argocd-agent-params`  with parameters related to agent.mode,agent.creds, agent.namespace, agent.server.address.
```
  agent.mode: autonomous
  agent.creds: mtls:any
  agent.tls.client.insecure: "false"
  agent.tls.root-ca-path: "/app/config/tls/ca.crt"
  agent.tls.client.cert-path: "/app/config/tls/tls.crt"
  agent.tls.client.key-path: "/app/config/tls/tls.key"
  agent.log.level: info
  agent.namespace: argocd
  agent.server.address: <argocd-principal-server>
  agent.server.port: 443
  agent.metrics.port: 8181
```

> Also Update RBAC, rolebinding/clusterrolebinding with argocd, if pod is facing rbac issues. 

Author: understood

# Argo CD Agent Hub and Spoke Cluster Setup Documentation Using Operator
This document outlines the process of setting up Argo CD Agent in a hub and spoke cluster architecture. This configuration allows a central Argo CD instance (the hub) to manage applications deployed across multiple remote Kubernetes clusters (the spokes) through the Argo CD-agent.
## Prerequisites
Before proceeding with the setup, ensure you have the following:

> A running OpenShift/kubernetes cluster designated as the hub, with Argo CD/OpenShift GitOps Operator installed.

Reviewer: ```suggestion
A running OpenShift/kubernetes cluster designated as the hub, with Argo CD/OpenShift GitOps Operator installed.
```
```suggestion
- A running OpenShift/kubernetes cluster designated as the hub, with Argo CD/OpenShift GitOps Operator installed.
```

- One or more Kubernetes clusters designated as spokes, Argo CD/OpenShift GitOps Operator installed.

> - oc configured to access both hub and spoke clusters.

Reviewer: ```suggestion
- `oc` binary configured to access both hub and spoke clusters.
```

- Operator must be installed in [cluster scope](https://argocd-operator.readthedocs.io/en/stable/usage/basics/#cluster-scoped-instance) mode.
- Apps in any Namespace must be enabled for Hub Cluster.

> - Argocd-agentctl binary (for non-production scenario)

Reviewer: ```suggestion
- `argocd-agentctl` binary (for non-production scenario)
```

    --ip <ip addresses of principal> \
    --dns <dns names of principal>
```

For the --ip and --dns values, you want to specify all addresses and DNS names that the principal will be reachable at for the agents, for example --ip 5.5.5.5 --dns my-principal.example.com. If there is a load balancer in front of the principal, make sure it will pass through traffic to the principal - otherwise, client certificates might not be used for authentication.

Issue the certificate for the principal's resource proxy:
The principal's resource proxy needs to be reachable by your Argo CD API server (argocd-server), which usually runs in the same cluster as the principal. Typically, this service should not be reachable outside the cluster and is exposed using a Kubernetes Service.

```
argocd-agentctl pki issue resource-proxy \
    --principal-context <control plane context> \
    --ip <ip addresses of resource proxy> \
    --dns <dns names of principal>
```

Deploy principal on hub cluster using argocd-operator/gitops-operator using Argo CD CR given below

```
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd
spec:
  controller:
    enabled: false
  argoCDAgent:

>     principal:

Reviewer 4: By default, the principal pod deployed by the operator [uses self-signed certs](https://github.com/argoproj-labs/argocd-operator/blob/ed071b55e5ce2d80ea25d2c9b57aae4ed50c342f/controllers/argocdagent/configmap.go#L309).  I think the TLS certs generated above using the `pki` command will not be used in that case. Should we set `ARGOCD_PRINCIPAL_TLS_SERVER_ALLOW_GENERATE` to false?

Author: I think it's okay to have it as it is. we can make an update to this part if we decide user need to update this to disable auto generated tls server.

      enabled: true
      allowedNamespaces: 
        - "*"
      jwtAllowGenerate: true
      auth: "mtls:CN=([^,]+)"
      logLevel: "trace"
      image: "<quay.io/user/argocd-agent:v1>"
  sourceNamespaces:
    - "agent-managed"
    - "agent-autonomous"  					
```

The above CR should create all the necessary resource for Argo CD as well as argocd-agent principal in argocd namespace.

Create argocd-redis secret, because principal looks for it to fetch redis authentication details.

```
oc create secret generic argocd-redis -n argocd --from-literal=auth="$(oc get secret argocd-redis-initial-password -n argocd -o jsonpath='{.data.admin\.password}' | base64 -d)"
```

## Setting up agent workload cluster 

### Configure Argo-CD for Agent 
 

> Argo CD instance on Agent cluster

Reviewer: This instance had o be made cluster scoped while testing

Reviewer: Its mentioned for the operator, not operand

> Creating Argo CD instance for Workload/spoke cluster.

Reviewer: I had to create an AppProject as well, during testing

```
apiVersion: argoproj.io/v1alpha1
kind: AppProject
metadata:
  name: default
  namespace: argocd
spec:
  clusterResourceWhitelist:
    - group: '*'
      kind: '*'
  destinations:
    - namespace: '*'
      server: '*'
  sourceRepos:
    - '*'
```

Author: I think agent does appProject sync, I can add this in the bottom of troubleshooting list

Reviewer: I'll find the error for troubleshooting section

```
  apiVersion: argoproj.io/v1beta1
  kind: ArgoCD
  metadata:
    name: argocd
  spec:
    server:
      enabled: false
```

### Configure Agent in managed mode

Before installing agent resources create 

> Create a TLS secret containing the issued certificate for agent

Reviewer: ```suggestion
a TLS secret containing the issued certificate for agent
```

Create the PKI on the agent:
Run this command while connected to principal
```

> argocd-agentctl pki issue agent <agent-name> --agent-context <workload context> --agent-namespace <workload namespace> --upsert

Reviewer: ```suggestion
argocd-agentctl pki issue agent <agent-name> --principal-context <principal context> --agent-context <workload context> --agent-namespace <workload namespace> --upsert

```

```
This should create all the required agent related resources.

Update the configMap with name `argocd-agent-params`  with parameters related to agent.mode,agent.creds, agent.namespace, agent.server.address.	
```
  agent.keep-alive-ping-interval: 50s
  agent.mode: managed
  agent.creds: mtls:any
  agent.tls.client.insecure: "false"
  agent.tls.root-ca-path: "/app/config/tls/ca.crt"
  agent.tls.client.cert-path: "/app/config/tls/tls.crt"
  agent.tls.client.key-path: "/app/config/tls/tls.key"
  agent.log.level: info
  agent.namespace: <workload-namespace>
  agent.server.address: <argocd-principal-server>
  agent.server.port: 443
  agent.metrics.port: 8181
  agent.healthz.port: "8002"
  agent.tls.root-ca-secret-name: argocd-agent-ca
  agent.tls.secret-name: argocd-agent-client-tls
```
Also Update RBAC, rolebinding/clusterrolebinding with `workload-namespace`, if pod is facing rbac issues. 

### Configure Agent in Autonomous mode

Before installing agent resources create 
Create a TLS secret containing the issued certificate for agent

Create the PKI on the agent:
Run this command while connected to principal
```

> argocd-agentctl pki issue agent <agent-name> --agent-context <workload context> --agent-namespace argocd --upsert

Reviewer: ```suggestion
argocd-agentctl pki issue agent <agent-name> --principal-context <principal context> --agent-context <workload context> --agent-namespace argocd --upsert
```

```
data:
  agent.keep-alive-ping-interval: 50s
  agent.tls.client.insecure: 'false' 
  agent.server.port: '443'
  agent.tls.root-ca-path: ''
  agent.tls.client.cert-path: ''
  agent.tls.root-ca-secret-name: argocd-agent-ca
  agent.tls.secret-name: argocd-agent-client-tls
  agent.mode: autonomous
  agent.log.level: info
  agent.namespace: argocd
  agent.server.address: <principal server address>
  agent.metrics.port: '8181'
  agent.tls.client.key-path: ''
  agent.healthz.port: '8002'
  agent.creds: 'mtls:any'
```

#### Troubleshooting 
___

1. If pod fails to come up with error
```
time="2025-07-30T14:58:33Z" level=warning msg="INSECURE: Not verifying remote TLS certificate"
time="2025-07-30T14:58:33Z" level=info msg="Loading client TLS certificate from secret argocd/argocd-agent-client-tls"
[FATAL]: Error creating remote: unable to read TLS client from secret: could not read TLS secret argocd/argocd-agent-client-tls: secrets "argocd-agent-client-tls" is forbidden: User "system:serviceaccount:argocd:argocd-agent-agent" cannot get resource "secrets" in API group "" in the namespace "argocd"
```

> update the ClusterRoleBinding to update the subject namespace to `workload-namespace`.

Reviewer: ```suggestion
update the ClusterRoleBinding to update the subject namespace to workload-namespace.
```

# Argo CD Agent Hub and Spoke Cluster Setup Documentation Using Operator
This document outlines the process of setting up Argo CD Agent in a hub and spoke cluster architecture. This configuration allows a central Argo CD instance (the hub) to manage applications deployed across multiple remote Kubernetes clusters (the spokes) through the Argo CD-agent.
## Prerequisites
Before proceeding with the setup, ensure you have the following:

- A running OpenShift/kubernetes cluster designated as the hub, with Argo CD/OpenShift GitOps Operator installed.
- One or more Kubernetes clusters designated as spokes, Argo CD/OpenShift GitOps Operator installed.

> - `oc` binary configured to access both hub and spoke clusters.

Author: This particular doc is under `OpenShift` so using `oc` as main command.
WDYT about the above comment @jannfis ?

> - `oc` binary configured to access both hub and spoke clusters.

Reviewer 5: You are right. If it's for OpenShift then `oc` is good. Thanks. 
FYI I don't have permission to resolve my comments so please feel free to resolve them.

- Operator must be installed in [cluster scope](https://argocd-operator.readthedocs.io/en/stable/usage/basics/#cluster-scoped-instance) mode.
- Apps in any Namespace must be enabled for Hub Cluster.

> - `argocd-agentctl` binary (for non-production scenario)

Author: I took this from the [getting-started](https://argocd-agent.readthedocs.io/latest/getting-started/#pki) doc.
@jannfis might be able to share more details on this.

> - `argocd-agentctl` binary (for non-production scenario)

Reviewer 5: ```
$ argocd-agentctl 
Usage:
   [flags]
   [command]

Configuration
  agent       Inspect and manage agent configuration
  jwt         Inspect and manage JWT signing keys
  pki         !!NON-PROD!! Inspect and manage the principal's PKI
```

All the `pki` subcommands are not safe but the inspect subcommands are safe.

## Hub Cluster Setup

Hub cluster will be installed in argocd namespace

> Before you install the principal's manifest, you will need to create three

Reviewer: ```suggestion
Before you install the principal's manifest, you will need to create
```

Issue the certificate for the principal's resource proxy:
The principal's resource proxy needs to be reachable by your Argo CD API server (argocd-server), which usually runs in the same cluster as the principal. Typically, this service should not be reachable outside the cluster and is exposed using a Kubernetes Service.

```
argocd-agentctl pki issue resource-proxy \
    --principal-context <control plane context> \
    --ip <ip addresses of resource proxy> \
    --dns <dns names of principal>
```

Deploy principal on hub cluster using argocd-operator/gitops-operator using Argo CD CR given below

```
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: argocd
spec:
  controller:
    enabled: false
  argoCDAgent:
    principal:
      enabled: true
      allowedNamespaces: 
        - "*"
      jwtAllowGenerate: true
      auth: "mtls:CN=([^,]+)"
      logLevel: "trace"
      image: "ghcr.io/argoproj-labs/argocd-agent/argocd-agent:latest"
  sourceNamespaces:
    - "agent-managed"
    - "agent-autonomous"  					
```

> The above CR should create all the necessary resource for Argo CD as well as argocd-agent principal in argocd namespace.

Reviewer 5: I had to modify the operator deployment to have the following env to properly consume the ca generated from the argocd-agentctl 
```
    - name: ARGOCD_PRINCIPAL_TLS_SERVER_ALLOW_GENERATE
      value: "false"
```

Author: Did you use the subscription to inject this as env var?

Reviewer 3: It's weird that you had to set `ARGOCD_PRINCIPAL_TLS_SERVER_ALLOW_GENERATE` to false, because that is supposed to be the default.

Reviewer 5: I think the operator set it to true by default if the env is not present so maybe it's a bug with the operator.
https://github.com/argoproj-labs/argocd-operator/blob/master/controllers/argocdagent/configmap.go#L305-L310

Reviewer 6: It is fixed in PR https://github.com/argoproj-labs/argocd-operator/pull/1849. Please update the doc.

Create argocd-redis secret, because principal looks for it to fetch redis authentication details.

```
oc create secret generic argocd-redis -n argocd --from-literal=auth="$(oc get secret argocd-redis-initial-password -n argocd -o jsonpath='{.data.admin\.password}' | base64 -d)"
```

## Setting up agent workload cluster 

### Configure Argo-CD for Agent 
 
Argo CD instance on Agent cluster

Creating Argo CD instance for Workload/spoke cluster.
```
  apiVersion: argoproj.io/v1beta1
  kind: ArgoCD
  metadata:
    name: argocd
  spec:
    server:

>       enabled: false

Reviewer 5: ```
  redis:
    enabled: true
```

For managed mode, the agent deployment is looking for the `argocd-redis` secret so I am assuming the redis is required.

```
    state:
      waiting:
        message: secret "argocd-redis" not found
        reason: CreateContainerConfigError
```

So the 

`kubectl create secret generic argocd-redis -n argocd --from-literal=auth="$(kubectl get secret argocd-redis-initial-password -n argocd -o jsonpath='{.data.admin\.password}' | base64 -d)"`

is needed as well afterwards.

Author: I think it will be required for both the Modes, I have added it for both the sections, managed and autonomous.
I have updated the command 
```
kubectl create secret generic argocd-redis -n <workload namespace> --from-literal=auth="$(kubectl get secret argocd-redis-initial-password -n <argocd-namespace> -o jsonpath='{.data.admin\.password}' | base64 -d)"
```
`workload-namespace` and `argocd-namespace` are mentioned as agent can be present in a different namespace than argocd.

Also, we will not need to enable redis explicitly, as it's absent in the mentioned argocd CR, it will be created with default values.

Reviewer 5: Right. Thanks.

```

### Configure Agent in managed mode

Before installing agent resources create 
- a TLS secret containing the issued certificate for agent

Create the PKI on the agent:
Run this command while connected to principal
```
argocd-agentctl pki issue agent <agent-name>  --principal-context <principal context> --agent-context <workload context> --agent-namespace <workload namespace> --upsert
```

Apply the installation manifests for Argo CD-agent agent
```

> oc apply -n $(workload-namespace) -k 'https://github.com/argoproj-labs/argocd-agent/install/kubernetes/agent?ref=main'

Reviewer 5: The rolebinding and clusterrolebinding is using `default` namespace. They need to be patch to use `argocd` namespace or the agent pod will run into issue. The setting of default seems intentional and not a bug based on the conversation here: https://github.com/argoproj-labs/argocd-agent/pull/428#discussion_r2168878846

> oc apply -n $(workload-namespace) -k 'https://github.com/argoproj-labs/argocd-agent/install/kubernetes/agent?ref=main'

Author: I have added that part, in troubleshooting section down below, I think I can add a note that if using namespace other than `default`, user will need to run patch command.

```
This should create all the required agent related resources.

Update the configMap with name `argocd-agent-params`  with parameters related to agent.mode,agent.creds, agent.namespace, agent.server.address.	
```
  agent.keep-alive-ping-interval: 50s
  agent.mode: managed
  agent.creds: mtls:any
  agent.tls.client.insecure: "false"
  agent.tls.root-ca-path: "/app/config/tls/ca.crt"
  agent.tls.client.cert-path: "/app/config/tls/tls.crt"

>   agent.tls.client.key-path: "/app/config/tls/tls.key"

Reviewer 5: Using the path doesn't seem to work because the secrets are not mounted in the deployment yaml properly.
The easiest way to set them via secret name is already done below in `agent.tls.root-ca-secret-name` and `agent.tls.secret-name`. I think these three path values should be removed.

Reviewer 5: Sounds good.

  agent.log.level: info
  agent.namespace: <workload-namespace>
  agent.server.address: <argocd-principal-server>
  agent.server.port: 443
  agent.metrics.port: 8181
  agent.healthz.port: "8002"
  agent.tls.root-ca-secret-name: argocd-agent-ca
  agent.tls.secret-name: argocd-agent-client-tls
```
Also Update RBAC, rolebinding/clusterrolebinding with `workload-namespace`, if pod is facing rbac issues. 

### Configure Agent in Autonomous mode

Before installing agent resources create 

> Create a TLS secret containing the issued certificate for agent

Reviewer: ```suggestion
a TLS secret containing the issued certificate for agent
```

  agent.tls.root-ca-secret-name: argocd-agent-ca
  agent.tls.secret-name: argocd-agent-client-tls
  agent.mode: autonomous
  agent.log.level: info
  agent.namespace: argocd
  agent.server.address: <principal server address>
  agent.metrics.port: '8181'
  agent.tls.client.key-path: ''
  agent.healthz.port: '8002'
  agent.creds: 'mtls:any'
```

#### Troubleshooting 
___

1. If pod fails to come up with error
```
time="2025-07-30T14:58:33Z" level=warning msg="INSECURE: Not verifying remote TLS certificate"
time="2025-07-30T14:58:33Z" level=info msg="Loading client TLS certificate from secret argocd/argocd-agent-client-tls"
[FATAL]: Error creating remote: unable to read TLS client from secret: could not read TLS secret argocd/argocd-agent-client-tls: secrets "argocd-agent-client-tls" is forbidden: User "system:serviceaccount:argocd:argocd-agent-agent" cannot get resource "secrets" in API group "" in the namespace "argocd"
```

update the ClusterRoleBinding to update the subject namespace to workload-namespace.

```
kubectl patch clusterrolebinding argocd-agent-agent --type='json' -p='[{"op": "replace", "path": "/subjects/0/namespace", "value": "<workload-namespace>"}]'
```

> 2. If appProject if not present in agent cluster, create using the below source

Reviewer: ```suggestion
2. If AppProject is not present in workload cluster, create one using the below manifest
```

> 2. If appProject if not present in agent cluster, create using the below source

Reviewer: Error: 

```
Unable to create application: app is not allowed in project "default", or the project does not exist
```

> 2. If appProject if not present in agent cluster, create using the below source

Reviewer 5: Yes, you are right. Thank you.
